component/metadata/metadata.go

```diff
@@ -0,0 +1,110 @@
+package metadata
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/grafana/agent/component"
+	_ "github.com/grafana/agent/component/all"
+	"github.com/grafana/agent/component/common/loki"
+	"github.com/grafana/agent/component/discovery"
+)
+
+type DataType string
+
+var (
+	// DataTypeTargets represents things that need to be scraped. These are used by multiple telemetry signals
+	// scraping components and often require special labels, e.g. __path__ label is required for scraping
+	// logs from files using loki.source.file component.
+	DataTypeTargets = DataType("Targets")
+
+	// DataTypeLokiLogs represent logs in Loki format
+	DataTypeLokiLogs = DataType("Loki Logs")
+
+	DataTypeOTELTelemetry     = DataType("OTEL Telemetry")
+	DataTypePromMetrics       = DataType("Prometheus Metrics")
+	DataTypePyroscopeProfiles = DataType("Pyroscope Profiles")
+)
+
+type Metadata struct {
+	Accepts []DataType
+	Outputs []DataType
+}
+
+func (m Metadata) Empty() bool {
+	return len(m.Accepts) == 0 && len(m.Outputs) == 0
+}
+
+func ForComponent(name string) (Metadata, error) {
+	reg, ok := component.Get(name)
+	if !ok {
+		return Metadata{}, fmt.Errorf("could not find component %q", name)
+	}
+	return inferMetadata(reg.Args, reg.Exports), nil
+}
+
+func inferMetadata(args component.Arguments, exports component.Exports) Metadata {
```

Comment:
Support for more types can be added in this function, which is also easy to test and make sure we don't get unexpected relationships.